src/cargo/core/shell.rs

```diff
@@ -519,6 +550,14 @@ fn supports_color(choice: anstream::ColorChoice) -> bool {
     }
 }
 
+fn supports_unicode(stream: &dyn IsTerminal) -> bool {
+    if !stream.is_terminal() || supports_unicode::supports_unicode() {
+        true
+    } else {
+        false
+    }
```

Comment:
nit: While I know if-else is more readable, should we collapse it?

```suggestion
    !stream.is_terminal() || supports_unicode::supports_unicode()
```